Docs / debugging

holmanb opened the pull request:
## Context
Followup to PR #4500.
- add degraded state documentation
- add content describing how to rerun cloud-init
- add content describing how to start debugging cloudinit, categorized by failure state
- relocates various bits of content to more appropriate locations

Fixes https://github.com/canonical/cloud-init/issues/4476. 

## Merge type
- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

TheRealFalcon commented on doc/rtd/howto/predeploy_testing.rst:
```diff
@@ -3,139 +3,69 @@
+    Making cloud-init run again may be destructive and must never be done on a production system. Artifacts such as ssh keys or passwords may be overwritten.
 
-The following command initialises a container with user data:
+Option #1: Remove the logs and cache then reboot
```
Think it's worth mentioning 
`cloud-init clean --logs; cloud-init init --local; cloud-init init; cloud-init modules --mode=config; cloud-init modules --mode=final; cloud-init status --wait --long`? It avoids a costly reboot so this is my default unless it's a quick-reboot like LXD.

holmanb replied:
> cloud-init status --wait --long

Why include the --wait command? To ensure that the reporter thread cleans up?

holmanb replied:
I guess that I could include this, however this section is intended to tell a user how to re-run ALL of cloud-init, which this will not do. I will add another section.

TheRealFalcon replied:
> Why include the --wait command?

It was just habit from waiting for cloud-init to run when connecting into an instance. But if we run each stage explicitly here, there's no need for the last status command.

TheRealFalcon commented on doc/rtd/reference/ubuntu_stable_release_updates.rst:
```diff
@@ -0,0 +1,49 @@
+.. _stable_release_updates:
+
+Stable Release Update (SRU)
```
```suggestion
Stable Release Updates (SRU)
```
Official documentation includes the s

TheRealFalcon commented on doc/rtd/reference/ubuntu_stable_release_updates.rst:
```diff
@@ -0,0 +1,49 @@
+.. _stable_release_updates:
```
I liked that we provided specific instructions on how to install proposed and then run a test on that proposed instance. At the very least I think we should link to a procedure to install a package from proposed (I'm assuming one exists somewhere on the internet?)

holmanb replied:
I added this information as a how to guide and a link on the page. Let me know if that works for you.

TheRealFalcon replied:
Yes, the new version is great!

TheRealFalcon commented on doc/rtd/howto/predeploy_testing.rst:
```diff
@@ -3,139 +3,69 @@
 
-Multipass
-=========
+.. _how_to_rerun_cloud-init:
```
I don't think tutorials should replace the multipass, lxd, and qemu sections here. Tutorials are for learning. How-tos are for getting things done. If I want to quickly spin up an LXD instance with cloud-init data, I'm not going to want to walk through a tutorial to figure out how to do it.

holmanb replied:
That's fair. However, I am also going to replace the cloud-localds command with genisoimage - it has broader adoption than `cloud-utils`, and cloud-localds is a thin wrapper around it.

TheRealFalcon commented on doc/rtd/howto/check_user_data.rst:
```diff
@@ -1,5 +1,7 @@
-======================
+.. _check_user_data_cloud_config:
+
+How to check user data cloud config
```
s/check/validate ? See my reasons above

TheRealFalcon commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,138 @@
+.. _failure_states:
+
+Failure states: recoverable errors vs non-recoverable errors
```
Maybe just `Failure states`? This is kind of a large section heading

TheRealFalcon commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,138 @@
+Failure states: recoverable errors vs non-recoverable errors
+============================================================
+
+critical failure
```
```suggestion
Critical failure
```

TheRealFalcon commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,138 @@
+within the 'error' key nested under the module-level keys: 'init-local',
+'init', 'modules-config', 'modules-final'.
+
+recoverable failure
```
```suggestion
Recoverable failure
```

TheRealFalcon commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,138 @@
+critical failure
+----------------
+If cloud-init is unable to complete, error messages will now
+additionally be visible in output of `cloud-init status --format json`
```
Mentioned elseware, but I think we should prefer `--long`

TheRealFalcon commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,138 @@
+
+Cloud-init error codes
+----------------------
+ 0 - success
```
These 3 lines are rendering as a single line

TheRealFalcon commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,138 @@
+Reported state
+--------------
+Cloud-init can report its internal state via the `status --format json`
+subcommand undert the `'extended_status'` key.
```
```suggestion
subcommand under the `'extended_status'` key.
```

TheRealFalcon commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,138 @@
+Cloud-init can report its internal state via the `status --format json`
+subcommand undert the `'extended_status'` key.
+
+$ cloud-init status --format json | jq .extended_status
```
This would be better rendered as a code block

TheRealFalcon commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,138 @@
+.. _failure_states:
```
I'm not seeing any of the code blocks being rendered in this file. Nothing gets rendered at all in their place.

TheRealFalcon commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,138 @@
+
+See the list of all possible states:
+
+.. code-block: shell-session
```
Having an inline list here but then appendices for the next two sections is a little odd, especially when this is the longest list. Personally, I prefer them all inline, but I'm fine either way as long as we are consistent.

TheRealFalcon reviewed (changes requested):
Overall, a really good change. Thanks!

I left some inline comments. Additionally:
The TOC is messed up. I'm seeing multiple elements of pages being rendered in the TOC for both `How-to guides` and `Explanation`. E.g.:
![image](https://github.com/canonical/cloud-init/assets/153674/1d02067c-e031-4379-add8-f60979f09e03)

Also, I would personally put 'How to debug cloud-init' into `How-to guides` rather than `Development`, because debugging is often required even if there's no cloud-init development happening.
I think I would also move `Internal Files: data` to reference. It might also be useful for debugging, and not just development.

holmanb commented:
> Also, I would personally put 'How to debug cloud-init' into How-to guides rather than Development, because debugging is often required even if there's no cloud-init development happening.

I agree, which is why it was linked from the How-to section as well. It is linked from both pages, which I guess was an odd choice. I can drop the development one, I guess, though I'm not sure that I am convinced that content being categorically "how to" makes it unfit for a link from the dev docs page. Thoughts?

TheRealFalcon commented:
Ah, I didn't catch that it was linked in both places. I see your point. I'm fine with it but would defer to @s-makin on that one.

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,108 @@
+   :ref:`cloud config<user_data_formats-cloud_config>`, make sure
+   to :ref:`check your user data cloud config<check_user_data_cloud_config>`
+
+2. Check for errors in `cloud-init status --format json`
```
I think that when I was writing this I expected to point them to other keys as well, but I did not, and at this point I don't think I will, so I agree.

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,108 @@
+2. Check for errors in `cloud-init status --format json`
+   - what is the value of the `'errors'` key?
+   - what is the value of the `'recoverable_errors'` key?
+
```
seems appropriate in both

holmanb commented on doc/rtd/howto/predeploy_testing.rst:
```diff
@@ -3,139 +3,69 @@
 
-When a user launches a Multipass VM, user data can be passed by adding the
-``--cloud-init`` flag and the appropriate YAML file containing the user data:
+Cloud-init runs on "first boot", so simply rebooting the system will
```
right, I meant to add context / reword (hence the hanging footnote above). Thanks for catching that.

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,108 @@
+.. _how_to_debug:
```
Thanks! I also assume other `cloud-init` devs probably have some other starting strategies that might be worth documenting for users. I don't want this page to become a dumping ground for one-off debugging hints - the general cases are preferred here in my opinion, but I bet that we can expand on this page a bit more - suggestions are welcome :)

holmanb replied:
Pinging some frequent contributors for extra eyes on this page: I'm soliciting feedback and ideas for this new user-facing debugging introduction.

@blackboxsw @aciba90 @dermotbradley @igalic @cjp256

aciba90 replied:
Another approach to access to instances without ssh access is to create a user with password access (using the user-data) and log in via the cloud serial port console. This only works if `cc_users_groups` was able to run.

holmanb commented on doc/rtd/howto/predeploy_testing.rst:
```diff
@@ -3,139 +3,69 @@
+If you are using :ref:`user data cloud-config<user_data_formats-cloud_config>`
+format, you might wish to rerun just a single configuration module.
+Cloud-init provides the ability to run a single module in isolation and
+separate from boot. This command is:
```
lol, thanks - formatting issue: the command was a few lines down

holmanb commented:
@TheRealFalcon Great review! I really appreciate it. I think I've addressed all of your comments, except for the TOC tree one (no clue on that one, but haven't really tried either) - enough to request re-review though, I think.

Additionally, a couple of new pages:
- a page describing how to test -proposed cloud-init packages
- instructions for cloud-init with libvirt
- now the "predeploy testing" page is split into 'how run cloud-init locally' and 'how to re-run cloud-init', since there is overlap, but they are fundamentally different, and the title "predeploy testing" is much less descriptive than the two new titles.

CC @s-makin I know you were taking a look at this PR as well, sorry for the churn mid-review. Thankfully I don't think much/any content has been removed since you started reviewing, but a couple of things shifted around since you started reviewing, and I added more content as well that will additionally want more eyes.

s-makin commented on doc/rtd/reference/ubuntu_stable_release_updates.rst:
```diff
@@ -0,0 +1,51 @@
+
+* :ref:`Install a pre-release version of cloud-init<ubuntu_test_pre_release>`
+  from the **-proposed** APT pocket (e.g., **jammy-proposed**).
+* Upgrade cloud-init and attempt a clean run of cloud-init to assert
+   that the new version works properly on the specific platform and Ubuntu
+   series.
```
This one rendered weirdly on the RTD site

s-makin commented on doc/rtd/reference/user_files.rst:
```diff
@@ -0,0 +1,48 @@
+.. _user_files:
+
+User Files: log and configuration
```
Can we avoid the use of "user files:" somehow?

holmanb replied:
now that this has its own page that is redundant

s-makin reviewed (commented):
I'm not sure I caught everything, but I gave it a good go! Thanks for doing all this, overall I agree with the changes and reorganisation. Mostly I found formatting inconsistencies, but there are a few things that confused me as well, so apologies in advance for the questions!

TheRealFalcon commented:
I don't think this should be labeled 23.4. It's not for anything that's in 23.4 specifically, correct? Doc updates will go live as soon as the PR is merged, so I don't think this PR needs to block release.

holmanb commented:
> I don't think this should be labeled 23.4. It's not for anything that's in 23.4 specifically, correct? Doc updates will go live as soon as the PR is merged, so I don't think this PR needs to block release.

Sure, nothing technically here that needs to block 23.4. This just documents the Werror stuff, which is released in 23.4, so I labeled it with 23.4 because I was hoping to get them released together so that users that read the release notes can go read the docs related to this new feature immediately.

holmanb commented on doc/rtd/development/internal_files.rst:
```diff
@@ -0,0 +1,46 @@
+.. _internal_files:
+
+Internal Files: data
```
There is actually an "external files" page: doc/rtd/reference/user_files.rst. The internal vs external distinction is really user vs non-user.

holmanb commented on doc/rtd/development/internal_files.rst:
```diff
@@ -0,0 +1,46 @@
+Internal Files: data
+********************
+
+Cloud-init uses the filesystem to store internal state. These files
```
> Not sure what is meant by "store internal state".

Yeah, this is a devdoc. Cloud-init stores state between stages of cloud-init running. This is an implementation detail and not a thing that a user should need to know about or understand.

> Also, which files does "these files" refer to? Just the ones below?

Correct

s-makin replied:
What I meant was that "store internal state" was ambiguous - the question in my mind as I was reading it was "store the internal state of what?". It's worth clarifying here that it's referring to cloud-init storing its own internal state and not something else.

holmanb replied:
good point, thanks - wording is now clarified

holmanb commented on doc/rtd/development/internal_files.rst:
```diff
@@ -0,0 +1,46 @@
+  whether any errors occurred.
+* :file:`status.json`:
+  JSON file showing the datasource used, a breakdown of all four modules,
+  whether any errors occurred, and the start and stop times.
```
internal stages

holmanb commented on doc/rtd/development/internal_files.rst:
```diff
@@ -0,0 +1,46 @@
+It also contains the :file:`datasource` file that contains the full information
+about which datasource was identified and used to set up the system.
+
+Finally, the :file:`boot-finished` file is the last thing that
```
That cloud-init finished, I think?

s-makin replied:
So that's all it does? It has no significance beyond cloud-init saying "I'm done!"?

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+
+Cloud-init error codes
+----------------------
+0 - success
```
> I get the impression this might also need some introductory text - without context I'm a bit confused about whether these are universal error codes that are applied to all failures, or if they're more specific than that and only apply in certain cases.

+1, will do

> We may also want to indent these as a quote block to avoid them all being shown on the same line. My preference would be to have them in a bulleted list, with the code (and meaning) on the bulleted line in monospaced font, then the corresponding text under that bullet point on the next line (to make it clear where it belongs).

I think this comment applies to a prior version of this change. Could you please let me know if the updated version looks better?

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+Critical failure
+----------------
+If cloud-init is unable to complete, error messages will now
+additionally be visible in output of `cloud-init status --long`
```
I think actually that this wording should be removed rather than enumerating other less-user-friendly interfaces

s-makin replied:
That would also work! :)

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+Recoverable failure
+-------------------
+In the case that cloud-init is able to complete yet something goes awry,
+the service will now return with exit code 2, and error messages will be
```
Documenting former state and current state would reduce the value of this documentation, I think, and would maybe have made more sense if former state was actually documented prior (it wasn't). I'll remove this wording.

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+  WARNING
+  DEPRECATED
+  ERROR
+  CRITICAL
```
I think I've answered the other ones, but not sure how to word an answer for this one:

> Do all warnings get reported, or can they be suppressed if a more serious one overlaps it?

All warnings are reported, but there is no "overlap".

holmanb replied:
More detail added (now in exported_errors.rst)

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+and was unable to recover. In this case, something is likely seriously
+wrong with the system, or cloud-init has experienced a serious bug.
+
+If cloud-init exits with exit code 2, cloud-init was able to complete
+gracefully, however something went wrong and the user should investigate.
```
"complete gracefully" means cloud-init was able to "finish, but not do what expected"

holmanb replied:
Lots more improvements on this page and the two new ones - hopefully this is now more clearly defined?

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+
+Exported errors: Aggregated errors
+----------------------------------
+When a recoverable error occurs, the internal cloud-init state
```
> We haven't defined anywhere what a "recoverable error" is 

See the "error codes" section

> and it doesn't match the section heading, either.

This section only applies to recoverable errors, not unrecoverable errors, should that be clarified maybe?

s-makin replied:
Looking back to where you pointed, I think the definition of a recoverable error is fine (although thinking about it, I'd maybe put subsubheadings for the recoverable/non-recoverable error definitions up there to call those sections out specifically). The primary confusion for me is how recoverable/non-recoverable errors relate to Exported and Aggregated errors (as in the headings) since this isn't immediately obvious.

holmanb replied:
The exported and aggregated errors content relates to recoverable errors. I've broken this content out into a separate page to make that more clear.

holmanb commented on doc/rtd/howto/check_user_data.rst:
```diff
@@ -1,5 +1,7 @@
+How to validate user data cloud config
+======================================
 
 Two of the most common issues with cloud config user data are:
```
> Also, I'd suggest including the file/path that we're discussing here for clarity.

There is no file/path. We're discussing a config format, and how that config gets delivered to cloud-init varies by each datasource.

> Also also, from what I understand the user data config first line should only have #cloud-config and nothing else

Hmm, I'd have to check the code but I thought the code was more like "must start with", not "must only contain". I'll update the wording

holmanb commented on doc/rtd/howto/check_user_data.rst:
```diff
@@ -23,22 +23,21 @@ Or, to test YAML in a file:
 Validation service
 ------------------
 
 Another option to is to use the self-hosted HTTP `validation service`_,
```
@aciba90 thoughts?

aciba90 replied:
We could include this as part of this PR, or I can create an extra one after this one is merged if required.

```suggestion
Another option is to use the self-hosted HTTP `validation service`_.
This service enables programmatic access to cloud-config validation.
Other programs / services / machines can reach, via HTTP requests, to that previously deployed service to validate their cloud-configs.
This removes the need for the requesting computer to have cloud-init installed to perform the validation,
```

holmanb replied:
thanks! added

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,142 @@
+   :ref:`tutorials<tutorial_index>`. Use LXD or Qemu locally to get a shell or
+   logs then debug with :ref:`these steps<did_not_do_the_thing>`.
+
+3. Try copying the image to your local system, mount the filesystem locally
+   and inspect (create a chroot in the image filesystem before attempting the
+   steps in other sections).
```
> When you say "copy the image", do you mean the one used in the previous step? If they don't follow the tutorial, what image should they use?

Not necessarily. One can copy any image being used to launch the instance. This might mean exporting the image from a cloud to copy it locally, or yes if using the tutorial one could copy the image used in that.

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+visible in the output of `cloud-init status --long` under the top
+level 'recoverable_errors' key as well as within the 'error' key nested
+under the module-level keys: 'init-local', 'init', 'modules-config',
+'modules-final'.
```
Let me know if the rewording is better

s-makin replied:
Much clearer, thank you!

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+---------------------------------
+The keys 'errors' and 'recoverable_errors' are also exported for each
+stage to allow attribution of recoverable and non-recoverable errors
+to their source.
```
see new wording

s-makin replied:
Much more straightforward, thank you!

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,142 @@
+How to debug cloud-init
+***********************
+
+There are several cloud-init failure modes that one may need to
```
yep, good call

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,142 @@
+     to this instance?
+   - What does the documentation say about the parts of the configuration that
+     relate to this error?
+
```
Hmmm, you're making me wonder if it wouldn't be better to break up the new failure states documentation into "failure states" and "exported errors". The reason that they are grouped together is that the code changes that this documentation stems from heavily affect both, and there is a hierarchical relationship: "exported errors" occur for failure states that are "recoverable errors".

That said, from a user's perspective, it seems awkward to put them together. Here, the "exported errors" part is irrelevant, since that information is already in the logs, but the "failure states" part is very relevant to understanding what types of errors they might see.

Thoughts?

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,142 @@
+   - How does the cloud-init error relate to the configuration provided
+     to this instance?
+   - What does the documentation say about the parts of the configuration that
+     relate to this error?
```
> Do all of the errors raised via the logs already have corresponding docs to explain how to fix them, or is this something we need to address as well?

No, because we dont't attempt to foresee all of the possible ways that a user might pass an invalid configuration.

This is an attempt to get the user to think about whether their configuration actually makes sense and get them to actually go read the docs (many users don't).

holmanb commented on doc/rtd/howto/rerun_cloud_init.rst:
```diff
@@ -0,0 +1,89 @@
+    Making cloud-init run again may be destructive and must never be done on a production system. Artifacts such as ssh keys or passwords may be overwritten.
+
+Option #1: Remove the logs and cache then reboot
+************************************************
```
> Let's add a pair of bullet points on line 19

Bullet points right before the headers looks a bit awkward to me. How does it look now?

aciba90 reviewed (commented):
Thanks, @holmanb, for the much better doc structure. I made a non-full review.

holmanb commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,193 @@
+====
+Qemu is a general purpose computer hardware emulator that is capable of
+running virtual machines with hardware acceleration while emulating the host's
+instruction set on native architectures (via kvm), as well as emulating
```
KVM is just a Linux kernel module, not its own project, so I don't know that that really makes sense. Will add the QEMU link though.

holmanb commented on doc/rtd/howto/ubuntu_test_prerelease.rst:
```diff
@@ -0,0 +1,50 @@
+===========================
+
+Afterthe cloud-init team creates an upstream release, cloud-init will
+be released in the -proposed APT repository for a
```
>  Afaiu, -proposed is a pocket rather than an APT repository.

People say that word sometimes but I have no clue why they do and I've never seen it defined. The [ubuntu wiki](https://wiki.ubuntu.com/Testing/EnableProposed) uses the word repository, apt's docs on debian.org call it a [distribution](https://wiki.debian.org/SourcesList#sources.list_format) and the apt man page calls it a [suite](https://manpages.debian.org/bookworm/apt/sources.list.5.en.html).

If you have any clue which is most authoritative I'll happily change it to whichever is most correct. It seems to me though that there is disagreement about what to call it, and "pocket" is not one of the things I can actually find defined anywhere.

aciba90 replied:
A couple of pointers quasi-defining pockets and suites:

- https://github.com/canonical/ubuntu-maintainers-handbook/blob/a5a386be1be1507fbd1dd8842699b0758042169c/PackageTests.md?plain=1#L327
- https://help.launchpad.net/Glossary

As I understand it, -proposed represents apt suites and/or launchpad series + pocket depending on the context.

holmanb replied:
Thanks for the definitions @aciba90. I wish those were easier to find than in some team-specific documentation, or in a miscellaneous glossary not related to APT.

I'm happy to go whatever we collectively think is best here, @s-makin and @aciba90. Thoughts?

s-makin replied:
I guess that's on me for assuming that because I saw the terminology in the UMH it must have been standard packaging terminology that was defined somewhere officially :upside_down_face: even Debian hasn't really defined it. It feels wrong to refer to internal documentation for this sort of thing, so let's stick with the original terminology you used, or use suites since that's at least defined in the apt manpages so that if the users need to refer to the manpages, the terminology will at least be consistent (I definitely don't want to point them to the Ubuntu wiki).

TheRealFalcon replied:
I want to make a stronger case for the "pocket" terminology. Repository is fairly generic term, and can refer to pretty much any place to get software. You could talk about the "Main" or "Universe" repository along with the "-updates" or "-proposed" repository, even though "Universe" and "proposed" aren't interchangeable. "Pocket" is more specific and lets you know exactly what is being referred to. You can't really find it in Debian docs (except the glossary) because it's ubuntu-specific.

Debian defines it here:
https://wiki.debian.org/Glossary#pocket

`The Archive Repository Pockets` section explains it well here:
https://ubuntu.com/blog/ubuntu-updates-releases-and-repositories-explained#The_archive_repository_pockets

Pocket is found 15 times on the SRU page, vs 1 for repository:
https://wiki.ubuntu.com/StableReleaseUpdates

There's more explanation of it here:
https://wiki.ubuntu.com/SecurityTeam/FAQ#Repositories_and_Updates

It's used in other ubuntu-specific tooling like here (search pocket):
https://manpages.ubuntu.com/manpages/jammy/man1/add-apt-repository.1.html

Since pocket is an established term that is more specific than repository, I think we should be preferring it.

holmanb commented on doc/rtd/howto/ubuntu_test_prerelease.rst:
```diff
@@ -0,0 +1,50 @@
+Afterthe cloud-init team creates an upstream release, cloud-init will
+be released in the -proposed APT repository for a
+:ref:`period of testing<sru_testing>` Users are encouraged to test their
+workloads on this proposed release so that bugs can be caught and fixed prior
```
No, this refers to the cloud-init release, not the repository / suite / distribution / pocket / whatever that it comes from.

s-makin replied:
Noted - it's confusing though, given that we were also talking about -proposed. Can we change this to something like "on this pending release" or something similar to avoid the ambiguity?

holmanb replied:
Agreed, thanks @s-makin!

holmanb commented on doc/rtd/howto/ubuntu_test_prerelease.rst:
```diff
@@ -0,0 +1,50 @@
+------------------------------------------
+.. code-block:: bash
+
+    apt-get install -qy cloud-init
```
not needed I'll remove it

holmanb commented on doc/rtd/howto/ubuntu_test_prerelease.rst:
```diff
@@ -0,0 +1,50 @@
+    apt-get install -qy cloud-init
+
+Remove the proposed repository
+------------------------------
+Do this to avoid unintentially installing other unreleased packages.
```
worth a mention, thanks

holmanb commented on doc/rtd/reference/performance_analysis.rst:
```diff
@@ -0,0 +1,111 @@
+
+The `cloud-init` command has an analysis sub-command, :command:`analyze`, which
+parses any :file:`cloud-init.log` file into formatted and sorted events. It
+allows for detailed analysis of the most costly cloud-init operations, and to
+determine the long-pole in cloud-init configuration and setup. These
```
I love it. I thought the wording here was odd too.

holmanb commented on doc/rtd/reference/ubuntu_stable_release_updates.rst:
```diff
@@ -0,0 +1,51 @@
+Stable Release Updates (SRU)
+****************************
+
+Once upstream cloud-init has cut a new release, the Ubuntu Server team
```
+1, we can do better

holmanb commented on doc/rtd/reference/user_files.rst:
```diff
@@ -0,0 +1,48 @@
+Cloud-init log files
+====================
+
+``Cloud-init`` uses two files to log to:
```
The third is a directory (which yes is technically a file). I'll add some context to that one.

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+  stable output!
+- Exported errors and recoverable errors may occur at different stages
+  since users may reorder configuration modules to run at different
+  stages via cloud.cfg.
```
fixed, I think

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,126 @@
+
+Failure states
+==============
+
```
> It's not ideal to have sequential headers with no text in between them

Agreed, fixed.

> Could we take a second look at the header structure to avoid that?

I've broken some of this content out into two new pages:
```
doc/rtd/explanation/exported_errors.rst
doc/rtd/howto/status.rst
```

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,129 @@
+
+See the list of all possible states:
+
+.. code-block:: shell-session
```
context added

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,129 @@
+Reported recoverable error messages are grouped by the level at which
+they are logged. Complete list of levels:
+
+.. code-block:: shell-session
```
context added (now in exported_errors.rst)

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,129 @@
+
+The json representation of cloud-init stages (in run order) is:
+
+.. code-block:: shell-session
```
agreed, thanks!

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,142 @@
+One of the more challenging scenarios to debug is when you don't have
+shell access to your instance. You have a few options:
+
+1. Acquire console log messages and check for errors.
```
see the updated wording

holmanb replied:
see updated wording

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,142 @@
+.. code-block::
+
+   systemctl status cloud-init-local.service cloud-init.service\
+        cloud-config.service cloud-final.service
```
good point, I added context

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,142 @@
+   :ref:`cloud config<user_data_formats-cloud_config>`, make sure
+   to :ref:`validate your user data cloud config<check_user_data_cloud_config>`
+
+2. Check for errors in `cloud-init status --long`
```
I switched to --long for all examples in this doc (more readable)

holmanb commented on doc/rtd/howto/rerun_cloud_init.rst:
```diff
@@ -0,0 +1,89 @@
+If you are using :ref:`user data cloud-config<user_data_formats-cloud_config>`
+format, you might wish to rerun just a single configuration module.
+Cloud-init provides the ability to run a single module in isolation and
+separate from boot. This command is:
```
Fixed

holmanb commented on doc/rtd/howto/rerun_cloud_init.rst:
```diff
@@ -0,0 +1,89 @@
+
+.. note::
+    Each cloud-config module has a module ``FREQUENCY`` configured: ``PER_INSTANCE``, ``PER_BOOT``, ``PER_ONCE`` or ``PER_ALWAYS``. When a module is run by cloud-init, it stores a semaphore file in :file:`/var/lib/cloud/instance/sem/config_<module_name>.<frequency>` which marks when the module last successfully ran. Presence of this semaphore file prevents a module from running again if it has al ...
+
```
Fixed

holmanb commented on doc/rtd/howto/rerun_cloud_init.rst:
```diff
@@ -0,0 +1,89 @@
+
+Option #1: Manually run cloud-init stages
+*****************************************
+Note that during normal boot of cloud-init, the init system orders these
```
thanks, fixed!

holmanb commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,193 @@
+
+Boot the image with the iso attached
+------------------------------------
+Pass the :file:`seed.img` with QEMU:
```
thanks, fixed!

holmanb commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,193 @@
+deploying it to the cloud. Fortunately, there are several different virtual
+machine (VM) and container tools that are ideal for this sort of local
+testing.
+
```
agreed, done!

holmanb commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,193 @@
+    $ lxc profile create dev-user-data
+    $ lxc profile set dev-user-data user.user-data - < cloud-init-config.yaml
+    $ lxc launch ubuntu-daily:bionic test-container -p default -p dev-user-data
+
```
Agreed, done!

holmanb commented on doc/rtd/howto/rerun_cloud_init.rst:
```diff
@@ -0,0 +1,89 @@
+
+Option #2: Reboot the instance
+******************************
+
```
Good call, done.

holmanb commented:
@s-makin: I couldn't respond inline to this comment, not sure why:

> In general I like the approach. I'm not sure what order they should be in, but I'd hesitate to put the more challenging scenario at the top.

Generally I agree.

> I'd recommend using whatever is the most common as the first one

However I think that this one probably _is_ the most common one, especially for first time users due to misconfiguration.

> or we could look at having a top-of-the-page toc that directs people to specific scenarios (depending on how many more we want to add)

I added a list at the top, let me know what you think.

holmanb commented on doc/rtd/howto/index.rst:
```diff
@@ -20,8 +20,8 @@ How do I...?
 
    Test cloud-init locally before deploying <predeploy_testing.rst>
```
fixed, those extra headings came from some incorrect headers at the bottom of doc/rtd/howto/rerun_cloud_init.rst

holmanb commented:
Thanks for the detailed review @s-makin. You've provided some really good feedback and questions that are helping drive the improvement of user understanding of this project.

I think that I've addressed all of your comments. I left a couple of them unresolved so that we can continue the conversation in them.

Since your review, I've added a couple of new pages broken out from the "failure states" one: exported_errors.rst and status.rst. Hopefully this helps make clearer the relationships between concepts and avoids the dreaded "wall of text" that our docs have been in the past.

holmanb commented:
> Thanks, @holmanb, for the much better doc structure. I made a non-full review.

@aciba90 thanks for your input and feedback, it's much appreciated.

s-makin commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,71 @@
+``cloud-init status --long`` within the ``'error'``.
+
+The same errors will also be located under the key nested under the
+keys that store information related to each
```
```suggestion
module-level keys that store information related to each
```

s-makin commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,71 @@
+module-level keys: ``init-local``, ``init``, ``modules-config``,
+``modules-final`` for the same error keys.
+
+See :ref:`here<exported_errors>` for to learn how to use cloud-init's
+exported errors.
```
```suggestion
See :ref:`this more detailed explanation <exported_errors>` to learn how to use cloud-init's
exported errors.
```

s-makin commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,71 @@
+See :ref:`here<reported_status>` for more information on cloud-init's status.
+
+Where to next?
+--------------
+See :ref:`here<how_to_debug>` for a detailed guide to debugging cloud-init.
```
```suggestion
--------------

See :ref:`here<how_to_debug>` for a detailed guide to debugging cloud-init.
```

s-makin commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,71 @@
+If cloud-init exits with exit code 2, cloud-init was able to complete
+gracefully, however something went wrong and the user should investigate.
+
+See :ref:`here<reported_status>` for more information on cloud-init's status.
```
```suggestion
See :ref:`this more detailed explanation <reported_status>` for more information on cloud-init's status.
```

s-makin commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,179 @@
+   - what is the value of the ``'extended_status'`` key?
+   - what is the value of the ``'boot_status_code'`` key?
+
+See :ref:`here<reported_status>` for more information on the status.
```
```suggestion
See :ref:`our reported status explanation<reported_status>` for more information.
```

s-makin replied:
(I'm suggesting changing all these because using "here" as linking text is not considered best accessibility practice)

s-makin commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,179 @@
+   - what is the value of the ``'errors'`` key?
+   - what is the value of the ``'recoverable_errors'`` key?
+
+See :ref:`here<exported_errors>` for more information on these exported errors.
```
```suggestion
See :ref:`our guide on exported errors<exported_errors>` for more information.
```

s-makin commented on doc/rtd/howto/rerun_cloud_init.rst:
```diff
@@ -0,0 +1,95 @@
+
+.. warning::
+
+    Making cloud-init run again may be destructive and must never be done on a production system. Artifacts such as ssh keys or passwords may be overwritten.
```
```suggestion
    Making cloud-init run again may be destructive and must never be done on a production system. Artefacts such as SSH keys or passwords may be overwritten.
```

s-makin commented on doc/rtd/howto/rerun_cloud_init.rst:
```diff
@@ -0,0 +1,95 @@
+
+Most cloud-init configuration is only applied to the system once. This means
+that simply rebooting the system will only re-run a subset of cloud-init.
+Cloud-init provides two different options for rerunning cloud-init for
```
```suggestion
Cloud-init provides two different options for re-running cloud-init for
```

s-makin commented on doc/rtd/howto/rerun_cloud_init.rst:
```diff
@@ -0,0 +1,95 @@
+------------------------------
+
+Note that during normal boot of cloud-init, the init system runs these
+stages at specific points during boot. This means that the running code
+manually after booting the system might will cause the code to interact with
```
```suggestion
stages at specific points during boot. This means that running the code
manually after booting the system may cause the code to interact with
```

s-makin commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,218 @@
+====
+
+`QEMU`_ is a general purpose computer hardware emulator that is capable of
+running virtual machines with hardware acceleration as well as emulating the
+instruction sets of different architectures than the host's that are different
+from the host that you are running on.
```
```suggestion
running virtual machines with hardware acceleration as well as emulating the
instruction sets of different architectures than those of the host you are running on.
```

s-makin commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,218 @@
+Download a cloud image
+----------------------
+
+Download a Ubuntu image to run:
```
```suggestion
Download an Ubuntu image to run:
```

s-makin commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,218 @@
+
+    wget https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img
+
+Boot the image with the iso attached
```
```suggestion
Boot the image with the ISO attached
```

s-makin commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,218 @@
+Boot the image with the iso attached
+------------------------------------
+
+Boot the cloud image witth our configuration, :file:`seed.img`, to QEMU:
```
```suggestion
Boot the cloud image with our configuration, :file:`seed.img`, to QEMU:
```

s-makin commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,218 @@
+The now-booted image will allow for login using the password provided above.
+
+For additional configuration, users can provide much more detailed
+configuration in the empty files network-config and meta-data.
```
```suggestion
configuration in the empty files :file:`network-config` and :file:`meta-data`.
```

s-makin commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,218 @@
+===
+
+`LXD`_ offers a streamlined user experience for using Linux system containers.
+With LXD.
+
+The following command initialises a container with user data:
```
```suggestion
With LXD, the following command initialises a container with user data:
```

s-makin commented on doc/rtd/howto/run_cloud_init_locally.rst:
```diff
@@ -0,0 +1,218 @@
+Libvirt
+=======
+
+`Libvirt`_ is tool for managing virtual machines and containers.
```
```suggestion
`Libvirt`_ is a tool for managing virtual machines and containers.
```

s-makin commented on doc/rtd/howto/status.rst:
```diff
@@ -0,0 +1,52 @@
+
+Cloud-init is also capable of reporting when cloud-init has not been
+able to complete the tasks described in a user configuration. If cloud-init
+has experienced issues while running the extended status will include the word
```
```suggestion
has experienced issues while running, the extended status will include the word
```

s-makin commented on doc/rtd/howto/ubuntu_test_prerelease.rst:
```diff
@@ -0,0 +1,67 @@
+   # you know what to do
+
+If issues are found during testing, please file a `new cloud-init bug`_ and
+drop a message in the `#cloud-init IRC channel`_.
```
```suggestion
leave a message in the `#cloud-init IRC channel`_.
```

s-makin commented on doc/rtd/howto/ubuntu_test_prerelease.rst:
```diff
@@ -0,0 +1,67 @@
+    rm -f /etc/apt/sources.list.d/proposed.list
+    apt update
+
+Remove artifacts and reboot
```
```suggestion
Remove artefacts and reboot
```

holmanb commented:
Thanks @s-makin for the re-review! I think I've addressed everything!

holmanb commented:
@TheRealFalcon Looks like your review is still in a "requested changes" state. I still need to apply / split out the changes in the comments in this PR into logical commits. I'm happy to do that before or after review. I haven't force pushed in case you wanted to compare from when you left off, although `15 files changed, 549 insertions(+), 212 deletions(-)` with 3 new files since then so maybe you don't mind if I do that?

TheRealFalcon commented:
@holmanb , structure however you'd like. I'll re-review this one from scratch.

holmanb commented:
@TheRealFalcon Ready for review :)

TheRealFalcon commented on doc/rtd/howto/ubuntu_test_prerelease.rst:
```diff
@@ -0,0 +1,67 @@
+
+.. code-block:: bash
+
+   # you know what to do
```
What is the point of this code block? If it's to say "insert your own test procedure here", that's not super obvious and I don't think we should keep it.

TheRealFalcon reviewed (changes requested):
Just what you wanted...more comments on the PR :wink: 

Since this PR is already very large, I'm ok saving most comments for the future. For the ones that I think really should get fixed before merging, I prepended with `fix:`

General note:
I don't love the use of jq in our docs. While it's very commonly used by a certain subset of developer, I wouldn't consider it to be a universal Linux tool nor does it come pre-installed on most distros. Is it that bad to show actual output? Could we add emphasis to what we're trying to highlight with styling?

Rant:
I've never seen the word `artefact` spelled that way before. I know our docs should use British English but that feels more wrong to me than marmite!

holmanb commented on doc/rtd/howto/debug_user_data.rst:
```diff
@@ -1,43 +1,65 @@
 
-Another option to is to use the self-hosted HTTP `validation service`_,
-refer to its documentation for more info.
+Another option to is to use the self-hosted HTTP `validation service`_.
```
Agreed, users are not the right target for this content.

I'd still like to see this showcased as a capability somewhere.

I propose that we could recommend that new clouds might want to do pre-launch validation over in `datasource_creation.rst`, with the content in this section as a reference implementation.

Thoughts?

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,182 @@
+1. Check the output of ``cloud-init status --long``
+
+   - what is the value of the ``'extended_status'`` key?
+   - what is the value of the ``'boot_status_code'`` key?
```
> What use is the boot_status_code key to the user? Would a user understand what enabled-by-generator or any other state means?

Yeah, I guess not. We should really start requiring that new additions to user commands get additions to user docs to describe what they actually do and mean.

I suppose I can document the meaning of this too, but I'll make that a separate PR. From [the commit message](https://github.com/canonical/cloud-init/commit/274b23ec18bebc4c77357e4a5db08c0bc5627e5a) that added this key:
```
The boot-status-code can be:
 - 'unknown': systemd generators and ds-identify haven't run yet to
              determine if cloud-init should be run during this boot
 - 'disabled-by-marker-file': /etc/cloud/cloud-init.disabled exists
              which prevents cloud-init from ever running
 - 'disabled-by-generator': systemd generator ran ds-identify and
              determined no applicable cloud-init datasources
 - 'disabled-by-kernel-cmdline': kernel cmdline contained
              cloud-init=disabled
 - 'enabled-by-kernel-cmdline': kernel cmdline contained
              cloud-init=enabled
 - 'enabled-by-generator': ds-identify detected possible cloud-init
              datasources
 - 'enabled-by-sysvinit': enabled by default in SysV init environment
 ```

holmanb commented on doc/rtd/explanation/failure_states.rst:
```diff
@@ -0,0 +1,74 @@
+.. _failure_states:
```
Could be, but while related they are really different concepts. One is about the states of cloud-init, the other is related to extracting information if cloud-init is in one of those states. I'd rather keep them separate for now, but I'll add a link to the page since it isn't defined here - no need to re-define.

holmanb commented on doc/rtd/howto/debugging.rst:
```diff
@@ -0,0 +1,182 @@
+
+.. _did_not_run:
+
+Cloud-init did not run
```
I think I fixed all of the rendering issues on this page.

holmanb commented:
> I don't love the use of jq in our docs. While it's very commonly used by a certain subset of developer, I wouldn't consider it to be a universal Linux tool nor does it come pre-installed on most distros. Is it that bad to show actual output? Could we add emphasis to what we're trying to highlight with styling?


That's fair. It gets really verbose especially on older series that still expect the `.v1.` to exist, but I'm fine with trying to stick to more commonly available tools as much as possible and not assuming familiarity and availability of non-core tools.

holmanb commented:
> I want to make a stronger case for the "pocket" terminology.

Thanks for that. The blog describes it the best I've seen it so far, no objections to this request.

TheRealFalcon reviewed (approved):
LGTM! We might have a cloud-init record with this one :smile:

holmanb commented:
> LGTM! We might have a cloud-init record with this one 😄

Thanks! I just squashed the last couple of commits and will land once tests pass.
